core/arch/arm/kernel/secure_partition.c

```diff
@@ -989,12 +1022,201 @@ const struct ts_ops sp_ops __weak __relrodata_unpaged("sp_ops") = {
+{
+	enum teecore_memtypes mtype = MEM_AREA_RAM_SEC;
+	struct sp_pkg_header *sp_pkg_hdr = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	tee_mm_entry_t *mm = NULL;
+	struct fip_sp *sp = NULL;
+	uint64_t sp_fdt_end = 0;
+	size_t sp_pkg_size = 0;
+	vaddr_t sp_pkg_va = 0;
+	size_t num_pages = 0;
+
+	if (!tee_pbuf_is_sec(sp_pkg_pa, SMALL_PAGE_SIZE))
+		return TEE_ERROR_GENERIC;
+
+	/* Map only the first page of SP package to parse the header */
+	mm = tee_mm_alloc(&tee_mm_sec_ddr, SMALL_PAGE_SIZE);
+	if (!mm)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	sp_pkg_va = tee_mm_get_smem(mm);
+
+	if (core_mmu_map_contiguous_pages(sp_pkg_va, sp_pkg_pa, 1, mtype)) {
+		res = TEE_ERROR_GENERIC;
+		goto err;
+	}
+
+	sp_pkg_hdr = (struct sp_pkg_header *)sp_pkg_va;
+
+	if (sp_pkg_hdr->magic != SP_PKG_HEADER_MAGIC) {
+		EMSG("Invalid SP package magic");
+		res = TEE_ERROR_BAD_FORMAT;
+		goto err_unmap;
+	}
+
+	if (sp_pkg_hdr->version != SP_PKG_HEADER_VERSION) {
+		EMSG("Invalid SP header version");
+		res = TEE_ERROR_BAD_FORMAT;
+		goto err_unmap;
+	}
+
+	if (ADD_OVERFLOW(sp_pkg_hdr->img_offset, sp_pkg_hdr->img_size,
+			 &sp_pkg_size)) {
+		EMSG("Invalid SP package size");
+		res = TEE_ERROR_BAD_FORMAT;
+		goto err_unmap;
+	}
+
+	if (ADD_OVERFLOW(sp_pkg_hdr->pm_offset, sp_pkg_hdr->pm_size,
+			 &sp_fdt_end) || sp_fdt_end > sp_pkg_hdr->img_offset) {
+		EMSG("Invalid SP manifest size");
+		res = TEE_ERROR_BAD_FORMAT;
+		goto err_unmap;
+	}
+
+	core_mmu_unmap_pages(sp_pkg_va, mm->size);
+	tee_mm_free(mm);
+
+	/* Map the whole package */
+	num_pages = ROUNDUP_DIV(sp_pkg_size, SMALL_PAGE_SIZE);
+
+	mm = tee_mm_alloc(&tee_mm_sec_ddr, sp_pkg_size);
+	if (!mm)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	sp_pkg_va = tee_mm_get_smem(mm);
+
+	if (core_mmu_map_contiguous_pages(sp_pkg_va, sp_pkg_pa, num_pages,
+					  mtype)) {
+		res = TEE_ERROR_GENERIC;
+		goto err;
+	}
+
+	sp_pkg_hdr = (struct sp_pkg_header *)tee_mm_get_smem(mm);
+
+	sp = calloc(1, sizeof(struct fip_sp));
+	if (!sp) {
+		res = TEE_ERROR_OUT_OF_MEMORY;
+		goto err_unmap;
+	}
+
+	memcpy(&sp->sp_img.image.uuid, sp_uuid, sizeof(*sp_uuid));
+	sp->sp_img.image.ts = (uint8_t *)(sp_pkg_va + sp_pkg_hdr->img_offset);
+	sp->sp_img.image.size = sp_pkg_hdr->img_size;
+	sp->sp_img.image.flags = 0;
+	sp->sp_img.fdt = (uint8_t *)(sp_pkg_va + sp_pkg_hdr->pm_offset);
+	sp->mm = mm;
+
+	STAILQ_INSERT_TAIL(&fip_sp_list, sp, link);
+
+	return TEE_SUCCESS;
+
+err_unmap:
+	core_mmu_unmap_pages(tee_mm_get_smem(mm), mm->size);
+err:
+	tee_mm_free(mm);
+
+	return res;
+}
+
+static TEE_Result fip_sp_map_all(void)
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint64_t sp_pkg_addr = 0;
+	const void *fdt = NULL;
+	TEE_UUID sp_uuid = {};
+	int sp_pkgs_node = 0;
+	int subnode = 0;
+	int root = 0;
+
+	fdt = get_external_dt();
+	if (!fdt) {
+		EMSG("No SPMC manifest found");
+		return TEE_ERROR_GENERIC;
+	}
+
+	root = fdt_path_offset(fdt, "/");
+	if (root < 0)
+		return TEE_ERROR_BAD_FORMAT;
+
+	if (fdt_node_check_compatible(fdt, root, "arm,ffa-core-manifest-1.0"))
```

Comment:
This change mandates the external DT embeds a compatible node. Is it on purpose?